### PR TITLE
[dbnode] bootstrap fails when blocks go out of retention

### PR DESCRIPTION
**Pull request by soundvibe**
**What this PR does / why we need it**:

In some cases, when dbnode bootstrap takes a very long time (e.g. longer than namespace retention period), bootstrap source returns an error `persistent fs index bootstrap invalid ranges to persist:`, forcing a bootstrap retry. 
This PR  tries to fix this issue by handling such cases and skipping building out of retention index segments.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

**Review by notbdu — APPROVED** [review]
LGTM

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go [hunk 1]
```diff
@@ -94,6 +94,14 @@ func PersistBootstrapIndexSegment(
 	retentionOpts := ns.Options().RetentionOptions()
 	nowFn := resultOpts.ClockOptions().NowFn()
 	earliestRetentionTime := retention.FlushTimeStart(retentionOpts, nowFn())
+
+	// If bootstrapping is taking more time than our retention period, we might end up in a situation
+	// when earliestRetentionTime is larger than out block end time. This means that the blocks
+	// got outdated during bootstrap so we just skip building index segments for them.
+	if earliestRetentionTime.After(blockEnd) {
+		return result.IndexBlock{}, fs.ErrOutOfRetentionClaim
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/persist.go:102]: This error (`ErrOutOfRetentionClaim`) is taken from `index_claims_manager.go`, its message is `out of retention index volume claim`. I'm not sure if it is appropriate to reuse it here as this code path is not related to index claims.

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go [hunk 2]
```diff
@@ -94,6 +94,14 @@ func PersistBootstrapIndexSegment(
 	retentionOpts := ns.Options().RetentionOptions()
 	nowFn := resultOpts.ClockOptions().NowFn()
 	earliestRetentionTime := retention.FlushTimeStart(retentionOpts, nowFn())
+
+	// If bootstrapping is taking more time than our retention period, we might end up in a situation
+	// when earliestRetentionTime is larger than out block end time. This means that the blocks
+	// got outdated during bootstrap so we just skip building index segments for them.
+	if earliestRetentionTime.After(blockEnd) {
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/persist.go:101]: Shouldn't this also be true when `earliestRetentionTime` equals `blockEnd`?

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go [hunk 3]
```diff
@@ -253,6 +261,14 @@ func BuildBootstrapIndexSegment(
...
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/persist.go:268]: Shouldn't this also be true when `earliestRetentionTime` equals `blockEnd`?

File: src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go [hunk 4]
```diff
@@ -526,131 +527,179 @@ func TestBootstrapIndexWithPersistPrefersPersistedIndexBlocks(t *testing.T) {
...
+			expectedOutOfRetentionBlocks: 2,
+		},
 	}
-	times := newTestBootstrapIndexTimes(timesOpts)
-	firstIndexBlockStart := times.start.Truncate(testIndexBlockSize)
-
-	writeTSDBGoodTaggedSeriesDataFiles(t, dir, testNs1ID, times.start)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			dir := createTempDir(t)
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go:565]: Might have been a good idea to extract the test implementation from test table loop into a dedicated private function. This would remove 2 levels of nesting and probably result in a much cleaner diff.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go [hunk 5]
```diff
@@ -958,7 +958,13 @@ func (s *peersSource) processReaders(
 			blockStart,
 			blockEnd,
 		)
-		if err != nil {
+		if errors.Is(err, fs.ErrOutOfRetentionClaim) {
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:961]: Any tests for `peers/source.go` changes?

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 1 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go:102]: The reason I decided to reuse this error is because the callers were already handling it in most of the places, introducing new error type would force them to implement additional checks. Also the newly introduced error would probably be very similar so it might get confusing. 
I would probably suggest to rename the existing error to something more general like `ErrOutOfRetention` so it could be used in different contexts more easily.

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 2 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go:101]: Well, this is the question about block's upper range inclusiveness. If we have a block from `06:00` to `10:00`, do we treat it like `[06:00,10:00]` or `[06:00, 10:00)`? I wasn't sure, so to me it was safer to allow indexing when `earlierRetentionTime` is equals `blockEnd` then to forbid it.

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 3 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go:268]: Well, this is the question about block's upper range inclusiveness. If we have a block from `06:00` to `10:00`, do we treat it like `[06:00,10:00]` or `[06:00, 10:00)`? I wasn't sure, so to me it was safer to allow indexing when `earlierRetentionTime` is equals `blockEnd` then to forbid it.

File: src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go (hunk 4 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go:565]: Could you point me an example where such extraction is done? 
To me, anonymous method makes valuable role here by accessing a test context so there is no need to pass it or declare it.

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 2 above)
Comment (linasm) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go:101]: `10:00` would belong to the next block, so the end has to be exclusive to avoid overlap: `[06:00, 10:00)`.

File: src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go (hunk 4 above)
Comment (linasm) [reply src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go:565]: Usually it is not that useful, but in this case the body is 140 lines long, and the resulting diff is really hard to digest.
All of the context that you would have to pass would be just `(t, test)`.

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 2 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go:101]: ok, so if the end is exclusive, then it needs additional equals check.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 5 above)
Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:961]: Added out of retention test for peers

File: src/dbnode/storage/bootstrap/bootstrapper/persist.go (hunk 2 above)
Comment (linasm) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go]: FYI can use `!blockEnd.After(earliestRetentionTime)` for that.

Comment (soundvibe) [reply src/dbnode/storage/bootstrap/bootstrapper/persist.go]: Not sure if that's more readable but it is probably more efficient.

**Review by vpranckaitis — APPROVED** [review]
